Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -281,4 +282,148 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
+    · rintro r hr
+      rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+      rintro p hp
+      exact hp (hc hr) (c r).2
+    · ext x : 1
+      simp only [Set.mem_iInter, SetLike.mem_coe]
+      fconstructor
+      · refine λ h ↦ h λ r n hr hn ↦ ?_
+        rw [mem_set_smul_submodule_def, mem_sInf]
+        exact λ p hp ↦ hp hr hn
+      · exact λ h _ ↦ h
+
+lemma mem_set_smul_submodule (x : M) [SMulCommClass R R N] :
+    x ∈ s • N ↔ ∃ (c : R →₀ N), (c.support : Set R) ⊆ s ∧ x = c.sum λ r m ↦ r • m := by
+  fconstructor
+  · intros h
+    rw [set_smul_submodule_eq_submodule_map] at h
+    obtain ⟨c, hc, rfl⟩ := h
+    exact ⟨c, hc, rfl⟩
+
+  · rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+    rintro ⟨c, hc1, rfl⟩ p hp
+    simp only [Finsupp.sum, AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid, SetLike.val_smul]
+    exact Submodule.sum_mem _ λ r hr ↦ hp (hc1 hr) (c _).2
+
+@[simp] lemma mem_empty_smul_submodule (x : M) : x ∈ (∅ : Set R) • N ↔ x = 0 := by
+  fconstructor
+  · intro hx
+    rw [mem_set_smul_submodule_def, Submodule.mem_sInf] at hx
+    exact hx ⊥ (λ r _ hr ↦ hr.elim)
+  · rintro rfl; exact Submodule.zero_mem _
+
+lemma singleton_set_smul_submodule_eq [SMulCommClass R R M] (r : R) :
+    ({r} : Set R) • N = (r • N : Submodule R M) := by
+  apply set_smul_submodule_eq_of_le
+  · rintro r m rfl hm; exact ⟨m, hm, rfl⟩
+  · rintro _ ⟨m, hm, rfl⟩
+    rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+    intro p hp; exact hp rfl hm
+
+lemma mem_singleton_set_smul_submodule [SMulCommClass R R M] (r : R) (x : M) :
+    x ∈ ({r} : Set R) • N ↔ ∃ (m : M), m ∈ N ∧ x = r • m := by
+  rw [singleton_set_smul_submodule_eq]
+  change x ∈ (r • N : Set M) ↔ _
+  rw [coe_pointwise_smul]
+  fconstructor <;>
+  · rintro ⟨m, hm, rfl⟩; exact ⟨m, hm, rfl⟩
+
+protected def pointwiseSetMulActionSubmodule [SMulCommClass R R M] :
+    MulAction (Set R) (Submodule R M) where
+  one_smul x := show {(1 : R)} • x = x from SetLike.ext fun m =>
+    (mem_singleton_set_smul_submodule _ _ _).trans ⟨by rintro ⟨_, h, rfl⟩; rwa [one_smul],
+      fun h => ⟨m, h, (one_smul _ _).symm⟩⟩
+  mul_smul s t x := le_antisymm
+    (set_smul_submodule_le _ _ _ <| by
+      rintro _ n ⟨a, b, ha, hb, rfl⟩ hn
+      rw [mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem1 := ha)
+      exact mem_set_smul_submodule_of_mem_mem (mem1 := hb) (mem2 := hn))
+    (set_smul_submodule_le _ _ _ <| by
+      rintro r m hr hm
+      have : SMulCommClass R R x := ⟨fun r s m => Subtype.ext <| show r • s • m.1 = s • r • m.1
+        from smul_comm _ _ _⟩
+      rw [mem_set_smul_submodule] at hm
+      obtain ⟨c, hc1, rfl⟩ := hm
+      delta Finsupp.sum
+      simp only [AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid, SetLike.val_smul]
+      rw [Finset.smul_sum]
+      refine Submodule.sum_mem _ ?_
+      intro r' hr'
+      rw [← mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem2 := (c _).2)
+      exact ⟨r, r', hr, hc1 hr', rfl⟩)
```

Comment:
Another justification of using `sInf` is this proof that `s \bu t \bu N = (s * t) \bu N` is simple when we know that set-smul-submodule is the smallest of something